src/object/cli_obj.c

```diff
@@ -2627,3 +2662,369 @@ dc_obj_query_key(tse_task_t *api_task)
+	struct obj_verify_one_rdg_unpack_arg	 unpack_arg;
+	daos_handle_t				 oh;
+	daos_obj_id_t				 oid;
+	unsigned int				 reps;
+	int					 leader;
+	int					 i;
+	int					 j;
+	int					 rc = 0;
+
+	oh = obj_ptr2hdl(obj);
+	oid = obj->cob_md.omd_id;
+	reps = obj_get_replicas(obj);
+	D_ALLOC_ARRAY(oeda, reps);
+	if (oeda == NULL) {
+		D_ERROR(DF_OID" no MEM for verify group, start %u, reps %u\n",
+			DP_OID(oid), start, reps);
+		return -DER_NOMEM;
+	}
+
+	for (i = 0; i < reps; i++) {
+		daos_unit_oid_t		shard_id = {
+			.id_pub		= oid,
+			.id_shard	= start + i,
+		};
+
+		daos_enum_dkeys_init_arg(&oeda[i], shard_id);
+	}
+
+	unpack_arg.oh		= oh;
+	unpack_arg.oid		= oid;
+	unpack_arg.epoch	= &epoch;
+
+	while (1) {
+
+leader:
+		leader = obj_grp_leader_get(obj, start, obj->cob_version);
+		if (leader < 0) {
+			D_ERROR(DF_OID" no valid shard, start %u, reps %u: "
+				"rc = %d.\n",
+				DP_OID(oid), start, reps, leader);
+			D_GOTO(out, rc = leader);
+		}
+
+		D_ASSERT(leader >= start && leader < start + reps);
+		leader -= start;
+
+		for (i = 0, j = leader; i < reps; i++, j = (j + 1) % reps) {
+			rc = daos_enum_dkeys_do_list(oh, &epoch, &oeda[j],
+					list_cb,
+					DIOF_TO_SPEC_SHARD | DIOF_FLUSH_DTX);
+
+			/* We used stale pool map, need to check new leader. */
+			if (rc == -DER_STALE) {
+				D_DEBUG(DB_TRACE,
+					DF_OID" (reps %d, start %u) Got "
+					"stale when enum shard (%u), retry\n",
+					DP_OID(oid), reps, start, start + j);
+				goto leader;
+			}
+
+			if (rc < 0) {
+				if (rc != -DER_NONEXIST)
+					D_GOTO(out, rc);
+
+				rc = 0;
+				oeda[j].lost_shard = 1;
+			} else {
+				oeda[j].lost_shard = 0;
+
+				if (daos_anchor_is_eof(&oeda[j].dkey_anchor))
+					oeda[j].eof = 1;
+			}
+
+			if (j == leader)
+				continue;
+
+			if (oeda[leader].lost_shard != oeda[j].lost_shard) {
+				D_INFO(DF_OID" (reps %d, inconsistent) "
+				       "shard %u %s, but shard %u %s\n",
+				       DP_OID(oid), reps, start + leader,
+				       oeda[leader].lost_shard ?
+				       "does not exist" : "does exist",
+				       start + j, oeda[j].lost_shard ?
+				       "does not exist" : "does exist");
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+
+			if (oeda[leader].eof != oeda[j].eof) {
+				D_INFO(DF_OID" (reps %d, inconsistent) "
+				       "shard %u %s, but shard %u %s\n",
+				       DP_OID(oid), reps, start + leader,
+				       oeda[leader].eof ? "end" : "not end",
+				       start + j,
+				       oeda[j].eof ? "not end" : "end");
+				D_GOTO(out, rc = -DER_MISMATCH);
+			}
+		}
+
+		if (oeda[leader].lost_shard)
+			break;
+
+		for (i = 0, j = leader; i < reps; i++, j = (j + 1) % reps) {
+			oeda[j].data_size = 0;
+			if (oeda[j].fetch_buf != NULL) {
+				D_FREE(oeda[j].fetch_buf);
+				oeda[j].fetch_buf = NULL;
+			}
+
+			if (oeda[j].num != 0) {
+				daos_enum_dkeys_prep_unpack(&oeda[j]);
+				unpack_arg.shard = start + j;
+				unpack_arg.arg = &oeda[j];
+				rc = obj_enum_unpack(&oeda[j].enum_arg,
```

Comment:
unpack is improper now, it does a lot more than "unpack", we could rename it to obj_enum_parse()